d-sandbox

<div style="text-align: center; line-height: 0; padding-top: 9px;">
  <img src="https://databricks.com/wp-content/uploads/2018/03/db-academy-rgb-1200px.png" alt="Databricks Learning" style="width: 600px; height: 163px">
</div>

# CSE Coding Assignment
## Instructions

- Please answer all questions
- You can use any language you wish (e.g. Python, Scala, SQL...)
- Several Markdown cells require completion. Please edit the Markdown cells to include your answer.
- Your final notebook should compile without errors when you click "Run All"

**Please do not publish questions. This is a confidential assignment.**

### Creating a Cluster

You will need to create a Databricks Cluster. More information on this process is available here: https://docs.databricks.com/user-guide/clusters/create.html

## Getting Started

**REQUIRED:** Run the following cells exactly as written to retrieve the necessary Coding Assignment Data Sets from Amazon S3.

In [0]:
%sh curl --remote-name-all 'https://files.training.databricks.com/assessments/cse-take-home/{covertype,kafka,treecover,u.data,u.item}.csv'


[1/5]: https://files.training.databricks.com/assessments/cse-take-home/covertype.csv --> covertype.csv
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed

  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
100   133  100   133    0     0   2770      0 --:--:-- --:--:-- --:--:--  2770

[2/5]: https://files.training.databricks.com/assessments/cse-take-home/kafka.csv --> kafka.csv

100  135k  100  135k    0     0  6459k      0 --:--:-- --:--:-- --:--:-- 6459k

[3/5]: https://files.training.databricks.com/assessments/cse-take-home/treecover.csv --> treecover.csv

100  642k  100  642k    0     0  17.4M      0 --:--:-- --:--:-- --:--:-- 17.4M

[4/5]: https://files.training.databricks.com/assessments/cse-take-home/u.data.csv --> u.data.csv

100 1932k  100 1932k    0     0  85.7M      0 --:--:-- --:--:-- --:--:-- 85.7M

[5/5]: https://files.training.databricks

In [0]:
dbutils.fs.cp("file:/databricks/driver/covertype.csv", "dbfs:/FileStore/tmp/covertype.csv")
dbutils.fs.cp("file:/databricks/driver/kafka.csv", "dbfs:/FileStore/tmp/kafka.csv")
dbutils.fs.cp("file:/databricks/driver/treecover.csv", "dbfs:/FileStore/tmp/treecover.csv")
dbutils.fs.cp("file:/databricks/driver/u.data.csv", "dbfs:/FileStore/tmp/u.data.csv")
dbutils.fs.cp("file:/databricks/driver/u.item.csv", "dbfs:/FileStore/tmp/u.item.csv")

Out[2]: True

## Part 1: Reading and Parsing Data

### Question 1:  Code Challenge - Load a CSV

- Load the CSV file at `dbfs:/FileStore/tmp/nl/treecover.csv` into a DataFrame.
- Use Apache Spark to read in the data, assigned to the variable `treeCoverDF`.
- Your method to get the CSV file into Databricks isn't graded. We are only concerned with how you use Spark to parse and load the actual data. 
- Please use the `inferSchema` option.

In [0]:
# YOUR CODE HERE

treeCoverDF = spark.read.format('csv').options(header='true', inferSchema='true').load('/FileStore/tmp/treecover.csv')

### Question 2:  Code Challenge - Print the Schema

Use Apache Spark to display the Schema of the `treeCoverDF` Dataframe.

In [0]:
# YOUR CODE HERE
treeCoverDF.printSchema()

root
 |-- Id: integer (nullable = true)
 |-- Elevation: integer (nullable = true)
 |-- Aspect: integer (nullable = true)
 |-- Slope: integer (nullable = true)
 |-- Horizontal_Distance_To_Hydrology: integer (nullable = true)
 |-- Vertical_Distance_To_Hydrology: integer (nullable = true)
 |-- Horizontal_Distance_To_Roadways: integer (nullable = true)
 |-- Horizontal_Distance_To_Fire_Points: integer (nullable = true)
 |-- Cover_Type: integer (nullable = true)
 |-- Soil_Type: integer (nullable = true)
 |-- Wilderness_Area: integer (nullable = true)
 |-- Hillshade: string (nullable = true)



### Question 3:  Code Challenge - Rows & Columns

Use Apache Spark to display the number of rows and columns in the DataFrame.

In [0]:
# YOUR CODE HERE
print('No of Rows ', treeCoverDF.count())

column_list = ','.join(treeCoverDF.columns)
print('Column Names : ', column_list)

No of Rows  15120
Column Names :  Id,Elevation,Aspect,Slope,Horizontal_Distance_To_Hydrology,Vertical_Distance_To_Hydrology,Horizontal_Distance_To_Roadways,Horizontal_Distance_To_Fire_Points,Cover_Type,Soil_Type,Wilderness_Area,Hillshade


#Part 2: Analysis

### Question 4:  Code Challenge - Summary Statistics for a Feature

Use Apache Spark to answer these questions about the `treeCoverDF` DataFrame:
- What is the range - minimum and maximum - of values for the feature `elevation`?
- What are the mean and standard deviation of the feature `elevation`?

In [0]:
# YOUR CODE HERE

from pyspark.sql import functions as f



print('Elevation Stats : ', treeCoverDF.agg(f.min(f.col('Elevation')).alias('Min_Elevation'),f.max(f.col('Elevation')).alias('Max_Elevation'),f.avg(f.col('Elevation')).alias('Avg_Elevation'),f.stddev(f.col('Elevation')).alias('StdDev_Elevation')).show())
#print('Max Elevation : ', treeCoverDF.agg(max('Elevation')).show())
#print('Mean Elevation : ', treeCoverDF.agg(mean('Elevation')).show())
#print('Std Deviation Elevation : ', treeCoverDF.agg(std('Elevation')).show())


treeCoverDF.select(['Elevation']).describe().show()

+-------------+-------------+------------------+------------------+
|Min_Elevation|Max_Elevation|     Avg_Elevation|  StdDev_Elevation|
+-------------+-------------+------------------+------------------+
|         1863|         3849|2749.3225529100528|417.67818734804985|
+-------------+-------------+------------------+------------------+

Elevation Stats :  None
+-------+------------------+
|summary|         Elevation|
+-------+------------------+
|  count|             15120|
|   mean|2749.3225529100528|
| stddev|417.67818734804985|
|    min|              1863|
|    max|              3849|
+-------+------------------+



### Answer #4:

- Min `elevation`: `1863`
- Max `elevation`: `3849`
- Mean `elevation`: `2749.32`
- Standard Deviation of `elevation`: `417.67`

### Question 5:  Code Challenge - Record Count

Use Apache Spark to answer the following question:
- How many entries in the dataset have an `elevation` greater than or equal to 2749.32 meters **AND** a `Cover_Type` of 1 or 2?

In [0]:
# YOUR CODE HERE
## & treeCoverDF['Cover_Type'].isin([1,2])
treeCoverDF.filter( (treeCoverDF['Elevation'] >= 2749.32) & (treeCoverDF['Cover_Type'].isin([1,2])) ).count()

Out[7]: 3883

### Question 6: Code Challenge - Compute a Percentage

Use Apache Spark to answer the following question:
- What percentage of entries with `Cover_Type` 1 or 2 have an `elevation` at or above 2749.32 meters?

In [0]:
# YOUR CODE HERE

cnt = treeCoverDF.filter( (treeCoverDF['Elevation'] >= 2749.32) & (treeCoverDF['Cover_Type'].isin([1,2])) ).count()
total_cnt = treeCoverDF.count()

print('% of Entries {:.6}'.format((cnt/total_cnt) * 100 ))

% of Entries 25.6812


### Question 7: Code Challenge - Visualize Feature Distribution

Use any [visualization tool available in the Databricks Runtime](https://docs.databricks.com/user-guide/visualizations/index.html) to generate the following visualization:

- a bar chart that helps visualize the distribution of different Wilderness Areas in our dataset

In [0]:
# YOUR CODE HERE


#treeCoverDF.select('Wilderness_Area').show()

display(treeCoverDF.select('Wilderness_Area').groupBy('Wilderness_Area').count().orderBy('Wilderness_Area'))

Wilderness_Area count 1 3597 2 499 3 6349 4 4675

### Question 8: Code Challenge - Visualize Average Elevation by Cover Type 

Use any [visualization tool available in the Databricks Runtime](https://docs.databricks.com/user-guide/visualizations/index.html) to generate the following visualization:

- a bar chart showing the average elevation of each cover type with string labels for cover type

**NOTE: you will need to match the integer values in the column `treeCoverDF.Cover_Type` to the string values in `dbfs:/FileStore/tmp/nl/covertype.csv` to retrieve the Cover Type Labels. It is recommended to use an Apache Spark join.**

In [0]:
# YOUR CODE HERE

treeCoverTypeDF = spark.read.format('csv').options(header='true', inferSchema='true').load('/FileStore/tmp/covertype.csv')

treeCoverTypeDF.printSchema()

treeCoverDFJoin=treeCoverDF.join(treeCoverTypeDF,treeCoverDF.Cover_Type ==  treeCoverTypeDF.cover_type_key,"inner")

display(treeCoverDFJoin.groupBy('cover_type_label').avg('Elevation').orderBy('cover_type_label'))

root
 |-- cover_type_key: integer (nullable = true)
 |-- cover_type_label: string (nullable = true)



cover_type_label avg(Elevation) Aspen 2786.801388888889 Cottonwood/Willow 2223.42037037037 Douglas-fir 2423.276851851852 Krummholz 3362.7699074074076 Lodgepole Pine 2922.5402777777776 Ponderosa Pine 2398.4231481481484 Spruce/Fir 3128.025925925926

#Part 3: Data Ingestion, Cleansing, and Transformations

## Instructions 

This is a multi-step, data pipeline question in which you need to achieve a few objectives to build a successful job.

### Data Sets

#### `u.data.csv`

- The full u data set, 100000 ratings by 943 users on 1682 items. 
- Each user has rated at least 20 movies.  
- Users and items are numbered consecutively from 1. 
- The data is randomly ordered. 
- This is a tab separated file consisting of four columns: 
   - user id 
   - movie id 
   - rating 
   - date (unix seconds since 1/1/1970 UTC)

#### Desired schema

- `user_id INTEGER`
- `movie_id INTEGER`
- `rating INTEGER`
- `date DATE `

#### `u.item.csv`

- This is a `|` separated file consisting of six columns:
   - movie id
   - movie title
   - release date
   - video release date
   - IMDb URL
   - genre
- movie ids in this file match movie ids in `u.data`.

#### Desired schema

- `movie_id INTEGER`
- `movie_title STRING`

### Question 9:  Code Challenge - Load DataFrames

Use Apache Spark to perform the following:
1. define the correct schemas for each Data Set to be imported as described above  
   **note:** 
      - for `u.data.csv`, `date` *must* be stored using `DateType` with the format `yyyy-MM-dd`
      - you may need to ingest `timestamp` data using `IntegerType`
      - be sure to drop unneccesary columns for `u.item.csv`
1. import the two files as DataFrames names `uDataDF` and `uItemDF` using the schemas you defined and these paths:
   - `dbfs:/FileStore/tmp/u.data.csv`
   - `dbfs:/FileStore/tmp/u.item.csv`
1. order the `uDataDF` DataFrame by the `date` column

**NOTE:** Please display the DataFrames, `uDataDF` and `uItemDF` after loading.

#### `uDataDF`

In [0]:
# YOUR CODE HERE
## https://sparkbyexamples.com/pyspark/pyspark-read-csv-file-into-dataframe/

from pyspark.sql.types import StructType,StructField, StringType, IntegerType,LongType,TimestampType,DateType
uDataDF_schema = StructType([StructField("user_id",IntegerType(),True),StructField("movie_id",IntegerType(),True),StructField("rating",IntegerType(),True),StructField("date",LongType(),True)])

uDataDF = spark.read.format('csv').options(header=True, delimiter='\t').schema(uDataDF_schema).load('/FileStore/tmp/u.data.csv')
uDataDF = uDataDF.withColumn("date", uDataDF["date"].cast(TimestampType()).cast(DateType()))

uDataDF.printSchema()

print("No of rows " , uDataDF.count())

uDataDF.show(10)

root
 |-- user_id: integer (nullable = true)
 |-- movie_id: integer (nullable = true)
 |-- rating: integer (nullable = true)
 |-- date: date (nullable = true)

No of rows  99999
+-------+--------+------+----------+
|user_id|movie_id|rating|      date|
+-------+--------+------+----------+
|    186|     302|     3|1998-04-04|
|     22|     377|     1|1997-11-07|
|    244|      51|     2|1997-11-27|
|    166|     346|     1|1998-02-02|
|    298|     474|     4|1998-01-07|
|    115|     265|     2|1997-12-03|
|    253|     465|     5|1998-04-03|
|    305|     451|     3|1998-02-01|
|      6|      86|     3|1997-12-31|
|     62|     257|     2|1997-11-12|
+-------+--------+------+----------+
only showing top 10 rows



#### `uItemDF`

In [0]:
# YOUR CODE HERE
## https://sparkbyexamples.com/pyspark/pyspark-read-csv-file-into-dataframe/
from pyspark.sql.types import StructType,StructField, StringType, IntegerType,LongType,TimestampType,DateType
uItemDF_schema = StructType([StructField("movie_id",IntegerType(),True),StructField("movie_title",StringType(),True),StructField("release_date",LongType(),True),StructField("video_release_date",LongType(),True),StructField("imbd_url",StringType(),True),StructField("genre",StringType(),True)])

uItemDF = spark.read.format('csv').options(header=True, delimiter='|').schema(uItemDF_schema).load('/FileStore/tmp/u.item.csv')
uItemDF = uItemDF.withColumn("release_date", uItemDF["release_date"].cast(TimestampType()).cast(DateType()))
uItemDF = uItemDF.withColumn("video_release_date", uItemDF["video_release_date"].cast(TimestampType()).cast(DateType()))

uItemDF.printSchema()

print("No of rows " , uItemDF.count())

uItemDF.show(10)

root
 |-- movie_id: integer (nullable = true)
 |-- movie_title: string (nullable = true)
 |-- release_date: date (nullable = true)
 |-- video_release_date: date (nullable = true)
 |-- imbd_url: string (nullable = true)
 |-- genre: string (nullable = true)

No of rows  1681
+--------+--------------------+------------+------------------+--------------------+-----+
|movie_id|         movie_title|release_date|video_release_date|            imbd_url|genre|
+--------+--------------------+------------+------------------+--------------------+-----+
|       2|    GoldenEye (1995)|        null|              null|http://us.imdb.co...|    0|
|       3|   Four Rooms (1995)|        null|              null|http://us.imdb.co...|    0|
|       4|   Get Shorty (1995)|        null|              null|http://us.imdb.co...|    0|
|       5|      Copycat (1995)|        null|              null|http://us.imdb.co...|    0|
|       6|Shanghai Triad (Y...|        null|              null|http://us.imdb.co...|    0

### Question 10:  Code Challenge - Perform a Join

Use Apache Spark to do the following:
- join `uDataDF` and `uItemDf` on `movie_id` as a new DataFrame called `uMovieDF`  
   **note:** make sure you do not create duplicate `movie_id` columns
   
**NOTE:** Please display the DataFrame `uMovieDF`.

In [0]:
# YOUR CODE HERE

uMovieDF = uDataDF.join(uItemDF,on=['movie_id'], how="inner").select(uDataDF["*"],uItemDF["movie_title"],uItemDF["release_date"],uItemDF["video_release_date"],uItemDF["imbd_url"],uItemDF["genre"])

uMovieDF.printSchema()

uMovieDF.show(10)

root
 |-- user_id: integer (nullable = true)
 |-- movie_id: integer (nullable = true)
 |-- rating: integer (nullable = true)
 |-- date: date (nullable = true)
 |-- movie_title: string (nullable = true)
 |-- release_date: date (nullable = true)
 |-- video_release_date: date (nullable = true)
 |-- imbd_url: string (nullable = true)
 |-- genre: string (nullable = true)

+-------+--------+------+----------+--------------------+------------+------------------+--------------------+-----+
|user_id|movie_id|rating|      date|         movie_title|release_date|video_release_date|            imbd_url|genre|
+-------+--------+------+----------+--------------------+------------+------------------+--------------------+-----+
|    186|     302|     3|1998-04-04|L.A. Confidential...|        null|              null|http://us.imdb.co...|    0|
|     22|     377|     1|1997-11-07| Heavyweights (1994)|        null|              null|http://us.imdb.co...|    0|
|    244|      51|     2|1997-11-27|Legends o

### Question 11:  Code Challenge - Perform an Aggregation

Use Apache Spark to do the following:
1. create an aggregate DataFrame, `aggDF` by
  1. extracting the year from the `date` (of the review)
  1. getting the average rating of each film per year as a column named `average_rating`
  1. ordering descending by year and average rating
1. write the resulting dataframe to a table named "movie_by_year_average_rating" in the Default database  
   **note:** use `mode(overwrite)` 

#### Desired Schema
The schema of you resulting DataFrame should be:
- `year INTEGER`
- `movie_title STRING`
- `average_rating DOUBLE`

**NOTE:** Please display the DataFrame `aggDF`.

In [0]:
# YOUR CODE HERE

from pyspark.sql.functions import year

aggDF = uMovieDF.select(year("date").alias('year'),uMovieDF['movie_title'],uMovieDF['rating']).groupBy(['year','movie_title']).agg(f.avg(f.col('rating')).alias('average_rating')).orderBy(['year','average_rating'],ascending=False)

aggDF.printSchema()

print(aggDF.show(10))

aggDF.write.mode("overwrite").saveAsTable("movie_by_year_average_rating")

root
 |-- year: integer (nullable = true)
 |-- movie_title: string (nullable = true)
 |-- average_rating: double (nullable = true)

+----+--------------------+--------------+
|year|         movie_title|average_rating|
+----+--------------------+--------------+
|1998|Marlene Dietrich:...|           5.0|
|1998|        Faust (1994)|           5.0|
|1998|Mina Tannenbaum (...|           5.0|
|1998|Great Day in Harl...|           5.0|
|1998|Entertaining Ange...|           5.0|
|1998|     Star Kid (1997)|           5.0|
|1998|The Deadly Cure (...|           5.0|
|1998|  Prefontaine (1997)|           5.0|
|1998|Visitors, The (Vi...|           5.0|
|1998|Year of the Horse...|          4.75|
+----+--------------------+--------------+
only showing top 10 rows

None


## Part 4: Fun with JSON

JSON values are typically passed by message brokers such as Kafka or Kinesis in a string encoding. When consumed by a Spark Structured Streaming application, this json must be converted into a nested object in order to be used.

Below is a list of json strings that represents how data might be passed from a message broker.

**Note:** Make sure to run the cell below to retrieve the sample data.

In [0]:
%python


sampleJson = [
 ('{"user":100, "ips" : ["191.168.192.101", "191.168.192.103", "191.168.192.96", "191.168.192.99"]}',), 
 ('{"user":101, "ips" : ["191.168.192.102", "191.168.192.105", "191.168.192.103", "191.168.192.107"]}',), 
 ('{"user":102, "ips" : ["191.168.192.105", "191.168.192.101", "191.168.192.105", "191.168.192.107"]}',), 
 ('{"user":103, "ips" : ["191.168.192.96", "191.168.192.100", "191.168.192.107", "191.168.192.101"]}',), 
 ('{"user":104, "ips" : ["191.168.192.99", "191.168.192.99", "191.168.192.102", "191.168.192.99"]}',), 
 ('{"user":105, "ips" : ["191.168.192.99", "191.168.192.99", "191.168.192.100", "191.168.192.96"]}',), 
]


### Question 12:  Code Challenge - Count the IPs

Use any coding techniques known to you to parse this list of JSON strings to answer the following question:
- how many occurrences of each IP address are in this list?

#### Desired Output
Your results should be this:


| ip | count |
|:-:|:-:|
| `191.168.192.96` | `3` |
| `191.168.192.99` | `6` |
| `191.168.192.100` | `2` |
| `191.168.192.101` | `3` |
| `191.168.192.102` | `2` |
| `191.168.192.103` | `2` |
| `191.168.192.105` | `3` |
| `191.168.192.107` | `3` |

**NOTE:** The order of your results is not important.

In [0]:
# YOUR CODE HERE
from pyspark.sql.functions import explode
## https://kontext.tech/column/spark/284/pyspark-convert-json-string-column-to-array-of-object-structtype-in-data-frame
sampleJSON2 = []
import ast
for item in sampleJson:
  item_dict = ast.literal_eval(item[0])
  data = [item_dict]
  sampleJSON2.append(data)

  
uipsDF = sqlContext.read.json(sc.parallelize(sampleJSON2))  
#uipsDF.show()
#uipsDF.printSchema()
uipDF = uipsDF.select(explode('ips').alias('ip'))
#uipDF.show()

uipAggDF = uipDF.select('ip').groupBy('ip').count()

uipAggDF.show()

+---------------+-----+
|             ip|count|
+---------------+-----+
| 191.168.192.96|    3|
|191.168.192.101|    3|
|191.168.192.103|    2|
| 191.168.192.99|    6|
|191.168.192.105|    3|
|191.168.192.107|    3|
|191.168.192.102|    2|
|191.168.192.100|    2|
+---------------+-----+



This is the end of the official test.

-sandbox
&copy; 2019 Databricks, Inc. All rights reserved.<br/>
Apache, Apache Spark, Spark and the Spark logo are trademarks of the <a href="http://www.apache.org/">Apache Software Foundation</a>.<br/>
<br/>
<a href="https://databricks.com/privacy-policy">Privacy Policy</a> | <a href="https://databricks.com/terms-of-use">Terms of Use</a> | <a href="http://help.databricks.com/">Support</a>